frontends/web/src/components/aopp/aopp.tsx

```diff
@@ -38,164 +38,127 @@ const Banner = ({ children }: TProps) => (
 );
 
-type State = {
-  accountCode: accountAPI.AccountCode;
-  aopp?: aoppAPI.Aopp;
-}
-
-type Props = TranslateProps;
+const setAccountCodeDefault = (aopp: aoppAPI.Aopp | undefined, setAccountCode: React.Dispatch<React.SetStateAction<string>>) => {
+  if (aopp === undefined || aopp.state !== 'choosing-account') {
+    return;
+  }
+  if (aopp.accounts.length) {
+    setAccountCode(aopp.accounts[0].code);
+  }
+};
 
 const domain = (callback: string): string => new URL(callback).host;
 
-class Aopp extends Component<Props, State> {
-  public readonly state: State = {
-    accountCode: '',
-    aopp: undefined,
-  };
-  private unsubscribe?: TUnsubscribe;
+export const Aopp = () => {
+  const { t } = useTranslation();
 
-  public componentDidMount() {
-    this.setAccountCodeDefault();
-    this.unsubscribe = aoppAPI.subscribeAOPP(aopp => this.updateAOPP(aopp));
-    aoppAPI.getAOPP().then(aopp => this.setState({ aopp }));
-  }
+  const unsubscribe = useRef<TUnsubscribe>();
 
-  public componentWillUnmount() {
-    if (this.unsubscribe) {
-      this.unsubscribe();
+  const [accountCode, setAccountCode] = useState<accountAPI.AccountCode>('');
+  const [aopp, setAopp] = useState<aoppAPI.Aopp>();
+  const [shouldUpdateAccountCodeDefault, setShouldUpdateAccountCodeDefault] = useState(false);
+
+  const updateAOPP = (new_aopp: aoppAPI.Aopp) => {
+    if (new_aopp?.state === 'choosing-account'
+      && (
+        aopp?.state !== 'choosing-account'
+        || !equal(new_aopp.accounts, aopp?.accounts)
+      )) {
+      setShouldUpdateAccountCodeDefault(true);
     }
-  }
+    setAopp(new_aopp);
+  };
 
-  private updateAOPP(aopp: aoppAPI.Aopp) {
-    let shouldUpdateAccountCodeDefault = false;
-    this.setState(currentState => {
-      if (aopp?.state === 'choosing-account'
-        && (
-          currentState.aopp?.state !== 'choosing-account'
-          || !equal(aopp.accounts, currentState.aopp?.accounts)
-        )
-      ) {
-        shouldUpdateAccountCodeDefault = true;
-      }
-      return { aopp };
-    }, () => { // callback when state did update
-      if (shouldUpdateAccountCodeDefault) {
-        this.setAccountCodeDefault();
+  useEffect(() => {
+    setAccountCodeDefault(aopp, setAccountCode);
+    unsubscribe.current = aoppAPI.subscribeAOPP(aopp => updateAOPP(aopp));
+    aoppAPI.getAOPP().then(aopp => setAopp(aopp));
+    return () => {
+      if (unsubscribe.current) {
+        unsubscribe.current();
       }
-    });
-  }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
 
-  private setAccountCodeDefault() {
-    const { aopp } = this.state;
-    if (aopp === undefined || aopp.state !== 'choosing-account') {
-      return;
+  useEffect(() => {
+    if (shouldUpdateAccountCodeDefault) {
+      setAccountCodeDefault(aopp, setAccountCode);
     }
-    if (aopp.accounts.length) {
-      this.setState({ accountCode: aopp.accounts[0].code });
-    }
-  }
+  }, [shouldUpdateAccountCodeDefault, aopp]);
```

Comment:
I don't see why shouldUpdateAccountCodeDefault is now a state, couldn't you remove the effect and use shouldUpdateAccountCodeDefault just internally in updateAOPP function as it was before?

Reply:
It's a bit subtle but I think it is correct to do it like this. Reason:

Before we used the callback function that can also be passed to `setState(newState, callback)` this function runs after the component rendered (see [this](https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/23e06b19a2e1d358d9363b0498ec14b20661c504/frontends/web/src/components/aopp/aopp.tsx#L81)).

We passed the boolean into this function, which means it will be _carried over_ to the new render, just like you can e.g. capture some variable when creating a go routine. 

The proper way to implement this in a functional component (run some function (the callback) after the render when certain state changed, is by using `useEffect` with a dependency on that state. Now we also want to check the boolean there, but if it was not state or a ref this value would get lost, unlike when we passed it to the callback directly. 

I guess it could be a `ref` though. Should I change it?

Reply:
Right, we do not need it. We get the `new_aopp` when the effect (now following your suggestion) re-runs because aopp changed and can `setAccountDefault(new_aopp, setAccountCode)` immediately.